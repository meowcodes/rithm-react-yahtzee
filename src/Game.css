.Game-reroll {
  margin-top: 1em;
  font-size: 1.5em;
  border-radius: 0.2em;
}

.Game-reroll:hover {
  background-color: pink;
}

.Game-flash {
  background-color: rebeccapurple;
  animation-name: colorChange;
  animation-duration: 1s;
  animation-direction: alternate;
  animation-iteration-count: infinite;
}

@keyframes colorChange {
  from {background-color: peachpuff;}
  to {background-color: rebeccapurple;}
}