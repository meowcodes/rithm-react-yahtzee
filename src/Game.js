import React, { Component } from 'react';
import Dice from './Dice';
import ScoreTable from './ScoreTable';
import './Game.css';

const NUM_DICE = 5;
const NUM_ROLLS = 3;

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      dice: this.initialRoll(),
      locked: Array(NUM_DICE).fill(false),
      rollsLeft: NUM_ROLLS,
      scores: {
        ones: undefined,
        twos: undefined,
        threes: undefined,
        fours: undefined,
        fives: undefined,
        sixes: undefined,
        threeOfKind: undefined,
        fourOfKind: undefined,
        fullHouse: undefined,
        smallStraight: undefined,
        largeStraight: undefined,
        yahtzee: undefined,
        chance: undefined
      }
    };
    this.roll = this.roll.bind(this);
    this.doScore = this.doScore.bind(this);
    this.toggleLocked = this.toggleLocked.bind(this);
  }

  initialRoll() {
    let arr = Array.from({ length: NUM_DICE })
    return arr.map(n => Math.ceil(Math.random() * 6));
  }

  roll(evt) {
    // roll dice whose indexes are in reroll
    this.setState(st => ({
      // generate random val if not locked
      dice: st.dice.map(
        (d, i) => st.locked[i] ? d : Math.ceil(Math.random() * 6)),
      // locks all dice when out of rolls
      locked: st.rollsLeft > 1 ? st.locked : Array(NUM_DICE).fill(true),
      rollsLeft: st.rollsLeft - 1,
    }));
  }

  toggleLocked(idx) {
    // if roll count is 0 
    // exit
    if (this.state.rollsLeft > 0 && this.state.dice[0]){
    // toggle whether idx is in locked or not
      this.setState(st => ({
        locked: [
          ...st.locked.slice(0, idx),
          !st.locked[idx],
          ...st.locked.slice(idx + 1)
        ],
      }))
    } 
  }

  doScore(rulename, ruleFn) {
    if(this.state.scores[rulename] === undefined && this.state.dice[0]){
      // evaluate this ruleFn with the dice and score this rulename
      this.setState(st => ({
        scores: { ...st.scores, [rulename]: ruleFn(this.state.dice) },
        rollsLeft: NUM_ROLLS,
        locked: Array(NUM_DICE).fill(false),
      }));
      this.roll();
    }
  }

  render() {
    return (
      <section>
        {/* bind with arrow function toggleLocked */}
        <Dice dice={this.state.dice} locked={this.state.locked} handleClick={this.toggleLocked} />
        <button
          className="Game-reroll"
          disabled={this.state.locked.every(x => x)}
          onClick={this.roll}>
          {this.state.rollsLeft} Rerolls Left
        </button>
        <ScoreTable doScore={this.doScore} scores={this.state.scores} />
      </section >
    );
  }
}

export default Game;