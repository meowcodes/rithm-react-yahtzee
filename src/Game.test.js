import React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Game from "./Game";

// smoke test
it("renders without crashing", function () {
    mount(<Game />);
});

let wrapper;

beforeEach(function(){
    wrapper = mount(<Game />);
})

// snapshot test
it("matches snapshot", function () {
    let serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
});

it("toggles lock when die is clicked", function () {
    let rollBtn = wrapper.find(".Game-reroll").first();
    rollBtn.simulate("click");

    let die = wrapper.find(".Die").first();
    die.simulate("click");

    expect(die.html()).toContain("Die-locked");
});

it("does not roll dice when out of rolls", function() {
    wrapper.setState({ rollsLeft: 1 });

    let btn = wrapper.find(".Game-reroll").first();
    btn.simulate("click");

    let die = wrapper.find(".Die").first();
    die.simulate("click");

    expect(wrapper.state().locked).toEqual([true, true, true, true, true]);
    expect(die.html()).toContain("Die-locked");
});

it("scores correctly", function() {
    wrapper.setState({ 
        dice: [1,1,1,2,3]
    });

    let scoreRow = wrapper.find(".RuleRow-name").first();
    scoreRow.simulate("click");

    expect(wrapper.state().scores.ones).toEqual(3);
});

it("does not allow re-scoring", function() {
    wrapper.setState({ 
        dice: [1,1,1,2,3],
        scores: { ones: 1 } 
    });

    let scoreRow = wrapper.find(".RuleRow-name").first();
    scoreRow.simulate("click");

    expect(wrapper.state().scores.ones).toEqual(1);
});

// test that box can't be clicked before initial roll
it("does not allow box click before initial roll", function() {
    let die = wrapper.find(".Die").first();
    die.simulate("click");

    expect(die.html()).not.toContain("Die-locked");
});

// test that list can't be clicked before initial roll
it("does not allow list click before initial roll", function() {
    let scoreRow = wrapper.find(".RuleRow-name").first();
    scoreRow.simulate("click");

    expect(wrapper.state().scores.ones).toEqual(undefined);
});