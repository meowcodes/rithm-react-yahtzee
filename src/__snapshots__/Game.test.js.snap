// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Game>
  <section>
    <Dice
      dice={
        Array [
          2,
          2,
          2,
          6,
          5,
        ]
      }
      handleClick={[Function]}
      locked={
        Array [
          false,
          false,
          false,
          false,
          false,
        ]
      }
    >
      <div
        className="Dice"
      >
        <Die
          handleClick={[Function]}
          idx={0}
          key="0"
          locked={false}
          val={2}
        >
          <button
            className="Die"
            onClick={[Function]}
          >
            2
          </button>
        </Die>
        <Die
          handleClick={[Function]}
          idx={1}
          key="1"
          locked={false}
          val={2}
        >
          <button
            className="Die"
            onClick={[Function]}
          >
            2
          </button>
        </Die>
        <Die
          handleClick={[Function]}
          idx={2}
          key="2"
          locked={false}
          val={2}
        >
          <button
            className="Die"
            onClick={[Function]}
          >
            2
          </button>
        </Die>
        <Die
          handleClick={[Function]}
          idx={3}
          key="3"
          locked={false}
          val={6}
        >
          <button
            className="Die"
            onClick={[Function]}
          >
            6
          </button>
        </Die>
        <Die
          handleClick={[Function]}
          idx={4}
          key="4"
          locked={false}
          val={5}
        >
          <button
            className="Die"
            onClick={[Function]}
          >
            5
          </button>
        </Die>
      </div>
    </Dice>
    <button
      className="Game-reroll"
      disabled={false}
      onClick={[Function]}
    >
      3
       Rerolls Left
    </button>
    <ScoreTable
      doScore={[Function]}
      scores={
        Object {
          "chance": undefined,
          "fives": undefined,
          "fourOfKind": undefined,
          "fours": undefined,
          "fullHouse": undefined,
          "largeStraight": undefined,
          "ones": undefined,
          "sixes": undefined,
          "smallStraight": undefined,
          "threeOfKind": undefined,
          "threes": undefined,
          "twos": undefined,
          "yahtzee": undefined,
        }
      }
    >
      <div
        className="ScoreTable"
      >
        <section
          className="ScoreTable-section"
        >
          <h2>
            Upper
          </h2>
          <table
            cellSpacing="0"
          >
            <tbody>
              <RuleRow
                doScore={[Function]}
                name="Ones"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Ones
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Twos"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Twos
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Threes"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Threes
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Fours"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Fours
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Fives"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Fives
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Sixes"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Sixes
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
            </tbody>
          </table>
        </section>
        <section
          className="ScoreTable-section ScoreTable-section-lower"
        >
          <h2>
            Lower
          </h2>
          <table
            cellSpacing="0"
          >
            <tbody>
              <RuleRow
                doScore={[Function]}
                name="Three of Kind"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Three of Kind
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Four of Kind"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Four of Kind
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Full House"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Full House
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Small Straight"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Small Straight
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Large Straight"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Large Straight
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Yahtzee"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Yahtzee
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
              <RuleRow
                doScore={[Function]}
                name="Chance"
              >
                <tr
                  className="RuleRow RuleRow-active"
                  onClick={[Function]}
                >
                  <td
                    className="RuleRow-name"
                  >
                    Chance
                  </td>
                  <td
                    className="RuleRow-score"
                  />
                </tr>
              </RuleRow>
            </tbody>
          </table>
        </section>
      </div>
    </ScoreTable>
  </section>
</Game>
`;
